### DISCUSSION TITLE
Clarification on session.close() behavior and reuse in SQLAlchemy

### DISCUSSION THREAD

**Discussion by R3gardless** [sqlalchemy/sqlalchemy#11696] [Usage Questions] [answered]
Hi Guys,

I'm currently studying the SQLAlchemy documentation and I have some questions regarding the **`session.close()`** method and its behavior.

According to the SQLAlchemy docs, **`session.close()`** acts more like a "reset" rather than a true "close". This has led me to some confusion when comparing it to how sessions are managed in FastAPI, where a new session is created per request using **`sessionmaker()`**.

My questions are as follows:

1. When using **`sessionmaker()`** to create a session, does SQLAlchemy reuse a closed session or does it always create a new session instance?

2. If SQLAlchemy does not reuse closed sessions, why is the session instance not simply deleted by default when the **`close()`** method is called, instead of being reset for memory utilization?

For context, here is an example of how FastAPI manages sessions:

```Python
from sqlalchemy.orm import sessionmaker
from fastapi import Depends

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()
```

I would appreciate any clarification on this matter.

Thanks

**Comment by CaselIT** [comment, marked as answer]
Hi

> When using sessionmaker() to create a session, does SQLAlchemy reuse a closed session or does it always create a new session instance?

It always creates a new insurance, it has no collection of created instances 

> If SQLAlchemy does not reuse closed sessions, why is the session instance not simply deleted by default when the close() method is called, instead of being reset for memory utilization?

Not sure what you mean by deleted. How do you delete a python object why have a reference to?

**Reply by R3gardless** [reply]
Apologize for my lack of knowledge of Python.
When `session.close()` called shouldn't reference disappear?

Thanks

**Reply by CaselIT** [reply]
How would this work? Example 

```py
session.close()
print(session) # what should session be here?
```

**Reply by R3gardless** [reply]
Ah, I understand now. Thank you very much for your kind assistance.

**Reply by CaselIT** [reply]
No problem. You can also set this flag https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.close_resets_only to false so that a session effectively becomes unusable after a close
